src/main/scala/com/fulcrumgenomics/rnaseq/AnnotateByIntervalList.scala

```diff
@@ -0,0 +1,246 @@
+import com.fulcrumgenomics.fasta.Converters.ToSAMSequenceDictionary
+import com.fulcrumgenomics.rnaseq.OverlapAssociationStrategy.{Duplicate, LeastSpecific, MostSpecific}
+import com.fulcrumgenomics.sopt.{arg, clp}
+import com.fulcrumgenomics.util.{Metric, ProgressLogger}
+import enumeratum.EnumEntry
+import htsjdk.samtools.SAMSequenceDictionary
+import htsjdk.samtools.util.{CoordMath, Interval, IntervalList, OverlapDetector}
+
+@clp(group = ClpGroups.Fastq, description =
+  """
+    |Get percentage of bases falling in each interval by name.
+    |A regular expression is used to classify the name of the intervals to be counted and how they should be grouped.
+    |
+    |For example for the following interval list:
+    |
+    |```
+    |chr1    981971  982021  +       utr
+    |chr1    982022  982065  +       intron
+    |chr1    982066  982117  -       CDS;1
+    |chr1    998963  999059  +       utr
+    |chr1    999060  999432  -       CDS;2
+    |chr1    999433  999526  +       intron
+    |chr1    999527  999613  -       CDS;3
+    |chr1    999614  999692  +       intron
+    |chr1    999693  999973  -       CDS;4
+    |```
+    |
+    |Using the the regex `([^;]*).*` would produce 3 separate interval counts CDS, utr and intron.
+    |
+    |Counts template bases (paired-end) or read bases (single end) that are fully enclosed within an interval.
+    |By default if a read is enclosed by multiple intervals it will be counted for both intervals. This behavior
+    |can be changed using a different `OverlapAssociationStrategy`
+  """)
+class AnnotateByIntervalList
+(
+  @arg(flag = 'i', doc = "Mapped BAM file.") val input: PathToBam,
+  @arg(flag = 'l', doc = "Path to intervals") val intervals: PathToIntervals,
+  @arg(flag = 'o', doc = "Path to the output metrics file") val output: FilePath,
+  @arg(flag = 'R', doc = "The regular expression to use to classify the name of the interval") val nameRegex: Option[String] = None,
+  @arg(flag = 'u', doc = "The name of the intervals not in the given set of intervals") val unmatched: String = "unmatched",
+  @arg(flag = 'm', doc = "The minimum mapping quality for a read to be included.") val minMappingQuality: Int = 20,
+  @arg(flag = 'q', doc = "The minimum base quality for a base to be included.") val minBaseQuality: Int = 0,
+  @arg(flag = 'a', doc = "The strategy to determine which interval overlapping templates should be associated with")
+  val overlapAssociationStrategy: OverlapAssociationStrategy = OverlapAssociationStrategy.Duplicate
+) extends FgBioTool with LazyLogging {
+
+  Io.assertReadable(this.input)
+  Io.assertReadable(this.intervals)
+  Io.assertCanWriteFile(this.output)
+
+  override def execute(): Unit = {
+    val progress = ProgressLogger(logger, verb = "Processed", noun = "loci", unit = 1000000)
+    val reader = SamSource(this.input)
+    val dict = reader.dict
+
+    val (baseCounter, uniqueBaseCounter, templateCounter, uniqueTemplateCounter, overlapDetector) = {
```

Comment:
Case class for this maybe?  Is there a threshold of tuple that we like to use for creating a higher ordered case class?